core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java

```diff
@@ -439,6 +442,28 @@ public void removeInstance(@Nonnull IdentityUser user, Long stackId, String inst
         }
     }
 
+    public void removeInstances(IdentityUser user, Long stackId, Set<String> instanceIds) {
+        Stack stack = get(stackId);
+        authorizationService.hasWritePermission(stack);
+        Map<String, Set<Long>> groupToDiscoveryPrivateIdMap = new HashMap<>();
+        for (String instanceId: instanceIds) {
+            InstanceMetaData metadata = instanceMetaDataRepository.findByInstanceId(stackId, instanceId);
+            if (metadata == null) {
+                throw new NotFoundException(String.format("Metadata for instance %s not found.", instanceId));
+            }
+            if (groupToDiscoveryPrivateIdMap.containsKey(metadata.getInstanceGroupName())) {
+                Set<Long> privateIds = groupToDiscoveryPrivateIdMap.get(metadata.getInstanceGroupName());
+                privateIds.add(metadata.getPrivateId());
+                groupToDiscoveryPrivateIdMap.put(metadata.getInstanceGroupName(), privateIds);
```

Comment:
Putting privateIds back is not necessary, because the same Set is already in the Map with the same key.